numpy/polynomial/polyutils.py

```diff
@@ -674,10 +676,30 @@ def _fit(vander_f, x, y, deg, rcond=None, full=False, w=None):
         msg = "The fit may be poorly conditioned"
         warnings.warn(msg, RankWarning, stacklevel=2)
 
-    if full:
-        return c, [resids, rank, s, rcond]
-    else:
+    if not full and not cov:
         return c
+ 
+    return_tuple = [c]
+
+    if full:
+        return_tuple.append([resids, rank, s, rcond])
+    if cov:
+        v_base = inv(np.dot(lhs, lhs.T))
```

Comment:
I'm not sure this is the most numerically stable way to get the inverse of the inner product.

Reply:
If A and B are invertible, AB is invertible too. The Vandemonde Matrix  (in this case lhs) is invertible (https://math.stackexchange.com/questions/426932/why-are-vandermonde-matrices-invertible, https://proofwiki.org/wiki/Inverse_of_Vandermonde_Matrix). Also, the tranpose of a invertible matrix is invertible too. So, always this result is going to be invertible.

Reply:
@lucasSaavedra123 I think @bashtage is referring to numerical stability. It may be that a matrix is technically invertible, but if it close to degenerate the numerical routines may fail. See, e.g. how the covariance matrix is calculated from the Jacobian in `scipy.optimize.curve_fit` using a singular value decomposition.:

https://github.com/scipy/scipy/blob/c1ed5ece8ffbf05356a22a8106affcd11bd3aee0/scipy/optimize/_minpack_py.py#L881-L889
```
cost = 2 * res.cost  # res.cost is half sum of squares!
popt = res.x

# Do Moore-Penrose inverse discarding zero singular values.
_, s, VT = svd(res.jac, full_matrices=False)
threshold = np.finfo(float).eps * max(res.jac.shape) * s[0]
s = s[s > threshold]
VT = VT[:s.size]
pcov = np.dot(VT.T / s**2, VT)
```

Relatedly or unrelatedly: what is holding this feature from being merged in? It would be nice to see this critical feature in `polynomial.polyfit` so that I can start using the modern recommended strategy for working with polynomials.

Reply:
The author seem to have abondoned the PR, or at least hasn't tried to incorporate any of the feedback.

Reply:
@bashtage I had no time to finish it. Next days I will incorporate all your feedback.

Reply:
In SciPy, the Jacobian Matrix is returned by their least-square implementation. But in this case we haven't such function or at least lsqt() function does not calculate it (as far i noticed). How is Jacobian Matrix efficiently computed with NumPy?

Reply:
Polynomial fits have the form $\mathbf{y} = \mathbf{A}\mathbf{b}$, where $\mathbf{A}$ is a [Vandermonde matrix](https://en.wikipedia.org/wiki/Vandermonde_matrix) and $\mathbf{b}$ is the parameter vector. The fit is linear in the parameters, so $\mathbf{J} = \mathbf{A}$. Let me know if that needs more explanation.

Reply:
Thank you for your answer! Very clear, @jsdodge. I wanted to be sure that I was picking the correct matrix and it seems to be that I did it but test failed. I will check that. Do you have an idea to test numerical stability of the function? Correct covariance matrix is not enough from my point of view.

Reply:
Yes, it should be the covariance matrix. What test failed? It's easy to get $A$ and $A^T$ mixed up. That would cause you to get the transpose of the covariance matrix. If not that, then it's also easy to get the scaling of things mixed up. 1/weights, 1/weights squared, chi squared, etc.

Reply:
I don't know if you can follow the exact routine I copied from scipy. In the current code it calculates $(LL^T)^{-1}$ (where L=`lhs`). The SVD says $L = U\Sigma V^T$ so $(LL^T)^{-1} = (U\Sigma V^T V \Sigma^T U^T)^{-1} = U\Sigma^{-2} U^T$ because $U$ and $V$ are orthogonal and $\Sigma$ is diagonal. So you need to use $U$ instead of $V$ in the singular value decomposition. Something like:

```
U, s, _= svd(lhs, full_matrices=False)

threshold = np.finfo(float).eps * max(lhs.shape) * s[0] 
mask = s > threshold
s = s[mask]
U = U[:, mask]
v_base = np.dot(U / s**2, U.T)
```

Reply:
Useful comment, @jagerber48. Thank You. This is solved.

Reply:
I was confused by this comment, so I checked the code and found that this function [defines](https://github.com/lucasSaavedra123/numpy/blob/2ada30dd4883649f7a9dd24ccfed750c2095c52e/numpy/polynomial/polyutils.py#L624C1-L648C1) the variable `lhs` to be the transpose of what I, at least, would normally associate with it.

```python
    if deg.ndim == 0:
        lmax = deg
        order = lmax + 1
        van = vander_f(x, lmax)
    else:
        deg = np.sort(deg)
        lmax = deg[-1]
        order = len(deg)
        van = vander_f(x, lmax)[:, deg]


    # set up the least squares matrices in transposed form
    lhs = van.T
    rhs = y.T
    if w is not None:
        w = np.asarray(w) + 0.0
        if w.ndim != 1:
            raise TypeError("expected 1D vector for w")
        if len(x) != len(w):
            raise TypeError("expected x and w to have same length")
        # apply weights. Don't use inplace operations as they
        # can cause problems with NA.
        lhs = lhs * w
        rhs = rhs * w
```

It is also inconsistent with the definition elsewhere in the [`numpy.lib.polynomial.polyfit`](https://github.com/lucasSaavedra123/numpy/blob/2ada30dd4883649f7a9dd24ccfed750c2095c52e/numpy/lib/polynomial.py#L648-L663).

```python
    # set up least squares equation for powers of x
    lhs = vander(x, order)
    rhs = y


    # apply weighting
    if w is not None:
        w = NX.asarray(w) + 0.0
        if w.ndim != 1:
            raise TypeError("expected a 1-d array for weights")
        if w.shape[0] != y.shape[0]:
            raise TypeError("expected w and y to have the same length")
        lhs *= w[:, NX.newaxis]
        if rhs.ndim == 2:
            rhs *= w[:, NX.newaxis]
        else:
            rhs *= w
```

I could be in the minority here, but I would recommend changing the definition of `lhs` in [numpy.polynomial.polyutils._fit](https://github.com/lucasSaavedra123/numpy/blob/2ada30dd4883649f7a9dd24ccfed750c2095c52e/numpy/polynomial/polyutils.py#L595) so that it equates `lhs` with $\mathbf{A}$ in $\mathbf{y} = \mathbf{A}\mathbf{b}$, as is done in [`numpy.lib.polynomial.polyfit`](https://github.com/lucasSaavedra123/numpy/blob/2ada30dd4883649f7a9dd24ccfed750c2095c52e/numpy/lib/polynomial.py#L453). These definitions may be mere conventions, but it's easier to avoid misunderstanding if we can agree on an appropriate convention.

Reply:
Right, I was just commenting about how to use singular value decomposition to find the inverse of something like $LL^T$. I think the linked `scipy` example would work for finding the inverse of $L^TL$. I can't weigh in on the convention choice until I read the code you posted above thoroughly and understand the choices being made. But I can weigh in that I do prefer conventions to be consistent, *especially* within one codebase.

That said, is this PR the place to try to get these conventions right? Or should this PR just get the `cov` feature working with minimal changes and then "convention" cleanup can be considered a non-output changing refactor handled in a different PR? (Not expressing any opinion one way or the other here, legitimately asking. I don't have experience with `numpy` development.)

Reply:
> 

This PR, as its title says, adds the `cov` parameter in `polyfit` function and that's all. @jagerber48 help me to understand how to make the inverse numerically stable (suggested by @jsdodge). I think that conventions should be discussed but It is not the topic of the PR.

Reply:
OK with me.

Reply:
Regarding the numerical stability, another potential improvement would be to do the computation in a rescaled domain, such as that currently returned by [`numpy.polynomial.Polynomial.fit`](https://numpy.org/doc/stable/reference/generated/numpy.polynomial.polynomial.Polynomial.fit.html), then convert that back to the original domain. This also could be done in a separate PR.

Reply:
I (or you, as you wish) open the issue and I/others do it in a separate PR. What you mention is another enhancement so I see no problem to do it.